include/miral/miral/decoration_adapter.h

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIRAL_DECORATION_DECORATION_ADAPTER_H
+#define MIRAL_DECORATION_DECORATION_ADAPTER_H
+
+#include "mir/geometry/forward.h"
+#include "mir_toolkit/common.h"
+
+#include "miral/decoration_basic_manager.h"
+#include "miral/decoration_window_state.h"
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <optional>
+
+
+
+namespace mir
+{
+namespace input
+{
+class CursorImages;
+}
+namespace shell
+{
+class SurfaceSpecification;
+namespace decoration
+{
+class Decoration;
+}
+}
+namespace scene
+{
+class Session;
+}
+namespace compositor
+{
+class BufferStream;
+}
+namespace graphics
+{
+class GraphicBufferAllocator;
+class Buffer;
+}
+}
+
+namespace msh = mir::shell;
+namespace msd = msh::decoration;
+namespace ms = mir::scene;
+namespace mc = mir::compositor;
+namespace mg = mir::graphics;
+namespace geometry = mir::geometry;
+
+struct MirEvent;
+
+namespace miral
+{
+namespace decoration
```

Comment:
I'm still not convinced a namespace is appropriate. (I would probably nest auxiliary types in `CustomDecoration`.